stdlib/Artifacts/test/runtests.jl

```diff
@@ -78,44 +80,50 @@ end
-            @test isdir(c_simple_dir)
-            c_simple_exe_path = joinpath(c_simple_dir, "bin", "c_simple$(exeext)")
-            @test isfile(c_simple_exe_path)
-
-            # Dynamically-computed slash-indexing:
-            generate_bin_path(pathsep) = "c_simple$(pathsep)bin$(pathsep)c_simple$(exeext)"
-            @test isfile(@artifact_str(generate_bin_path("/")))
-            @test isfile(@artifact_str(generate_bin_path("\\")))
-        end
+    tempdir = joinpath(@__DIR__, "artifacts")
+    with_artifacts_directory(tempdir) do
+        exeext = Sys.iswindows() ? ".exe" : ""
+
+        # simple lookup, gives us the directory for `c_simple` for the current architecture
+        c_simple_dir = artifact"c_simple"
+        @test isdir(c_simple_dir)
+        c_simple_exe_path = joinpath(c_simple_dir, "bin", "c_simple$(exeext)")
+        @test isfile(c_simple_exe_path)
+
+        # Simple slash-indexed lookup
+        c_simple_bin_path = artifact"c_simple/bin"
+        @test isdir(c_simple_bin_path)
+        # Test that forward and backward slash are equivalent
+        @test artifact"c_simple\\bin" == artifact"c_simple/bin"
+
+        # Dynamically-computed lookup; not done at compile-time
+        generate_artifact_name() = "c_simple"
+        c_simple_dir = @artifact_str(generate_artifact_name())
+        @test isdir(c_simple_dir)
+        c_simple_exe_path = joinpath(c_simple_dir, "bin", "c_simple$(exeext)")
+        @test isfile(c_simple_exe_path)
+
+        # Dynamically-computed slash-indexing:
+        generate_bin_path(pathsep) = "c_simple$(pathsep)bin$(pathsep)c_simple$(exeext)"
+        @test isfile(@artifact_str(generate_bin_path("/")))
+        @test isfile(@artifact_str(generate_bin_path("\\")))
     end
 end
 
 @testset "@artifact_str Platform passing" begin
-    mktempdir() do tempdir
-        with_artifacts_directory(tempdir) do
-            win64 = Platform("x86_64", "windows")
-            mac64 = Platform("x86_64", "macos")
-            @test basename(@artifact_str("c_simple", win64)) == "444cecb70ff39e8961dd33e230e151775d959f37"
-            @test basename(@artifact_str("c_simple", mac64)) == "7ba74e239348ea6c060f994c083260be3abe3095"
-        end
+    tempdir = joinpath(@__DIR__, "artifacts")
+    with_artifacts_directory(tempdir) do
+        win64 = Platform("x86_64", "windows")
+        mac64 = Platform("x86_64", "macos")
+        @test basename(@artifact_str("c_simple", win64)) == "444cecb70ff39e8961dd33e230e151775d959f37"
+        @test basename(@artifact_str("c_simple", mac64)) == "7ba74e239348ea6c060f994c083260be3abe3095"
     end
-end
\ No newline at end of file
+end
+
+@testset "@artifact_str lazy" begin
+    tempdir = joinpath(@__DIR__, "artifacts")
+    with_artifacts_directory(tempdir) do
+        ex = @test_throws ErrorException artifact"socrates"
+        @test startswith(ex.value.msg, "Artifact \"socrates\" was not installed correctly.")
```

Comment:
We should test both the "not installed correctly" branch and the "must import `Pkg`" branch, and the branch that actually succeeds in downloading the lazy artifact.

Reply:
Yeah, I'm only testing the new code here, but someone should write/have written a test for the lazy code too